tests/libblight_protocol/test.c

```diff
@@ -0,0 +1,85 @@
+#include <libblight_protocol.h>
+#include <assert.h>
+#include <stdio.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <string.h>
+
+void test_c(){
+    fprintf(stderr, "Running C tests...\n");
+
+    fprintf(stderr, "Testing blight_header_from_data\n");
+    blight_header_t x;
+    x.type = 1;
+    x.ackid = 1;
+    x.size = 1;
+    blight_header_t header = blight_header_from_data((blight_data_t)&x);
+    assert(header.type == 1);
+    assert(header.ackid == 1);
+    assert(header.size == 1);
+    x.type = 0;
+    x.ackid = 0;
+    x.size = 0;
+    blight_header_t header2 = blight_header_from_data((blight_data_t)&x);
+    assert(header2.type == 0);
+    assert(header2.ackid == 0);
+    assert(header2.size == 0);
+
+    fprintf(stderr, "Testing blight_message_from_data\n");
+    blight_data_t data = malloc(sizeof(blight_header_t) + 1);
+    memcpy(data, &header, sizeof(blight_header_t));
+    data[sizeof(blight_header_t)] = 'a';
+    blight_message_t* message = blight_message_from_data(data);
+    assert(message->header.type == 1);
+    assert(message->header.ackid == 1);
+    assert(message->header.size == 1);
+    assert(message->data != NULL);
+    assert(*message->data == 'a');
+    blight_message_deref(message);
+    free(data);
+
+    fprintf(stderr, "Testing blight_bus_connect_system\n");
+    blight_bus* bus = NULL;
+    assert(blight_bus_connect_system(&bus) > 0);
+    assert(bus != NULL);
+
+    fprintf(stderr, "Testing blight_service_available\n");
+    assert(blight_service_available(bus));
+
+    fprintf(stderr, "Testing blight_service_open\n");
+    int res = blight_service_open(bus);
+    assert(res > 0);
+    close(res);
+
+    fprintf(stderr, "Testing blight_service_input_open\n");
+    res = blight_service_input_open(bus);
+    assert(res > 0);
+    int fd = res;
+    blight_bus_deref(bus);
+
+    fprintf(stderr, "Testing blight_message_from_socket\n");
+    res = -EAGAIN;
+    while(-res == EAGAIN || -res == EINTR){
+        res = blight_message_from_socket(fd, &message);
```

Comment:
@coderabbitai do you know why this would never return any data, despite the data being sent through the socket?

Reply:
The service does already send a ping request, and will continue sending pings on a timer. The code here waits forever for any message to arrive on the socket, but it never returns.

Reply:
Good catch on the `MSG_WAITALL | MSG_DONTWAIT`. I likely just appended `MSG_DONTWAIT` without thinking about it.

For point 2, as the outer code is a while loop running forever if there is no data, the 250ms timeout wouldn't matter.

For point 3, `blight_service_input_open` just dups a file descriptor passed via dbus to the application. The server side opens it with the following code (found in applications/display-server/connection.cpp):
```
if(::socketpair(AF_UNIX, SOCK_STREAM | SOCK_NONBLOCK, 0, fds) == -1){
    C_WARNING("Unable to open socket pair:" << strerror(errno));
}
m_clientFd = fds[0];
```

You can see that it's being set to non-block.

Reply:
After fixing the flags so that recv only uses `MSG_DONTWAIT` for it's calls to `::recv`, this issue still persists.

I have other applications that use the recv message and do not have this issue.

Reply:
That was it. I can't believe I didn't see it, and that a bot found it for me.